File: r-package/tests/testthat/test-discrete.R
test_that("int returns integer values in [min,max]", {
  rng <- randompack_rng("pcg64")
  x <- rng$int(10, -2, 3)
  expect_type(x, "integer")
  expect_length(x, 10)
  expect_true(all(x >= -2))
  expect_true(all(x <= 3))
})

test_that("perm returns a permutation of 1..n", {
  rng <- randompack_rng("pcg64")
  x <- rng$perm(10)
  expect_type(x, "integer")
  expect_length(x, 10)
  expect_identical(sort(x), 1:10)
})

test_that("sample returns unique values in 1..n", {
  rng <- randompack_rng("pcg64")
  x <- rng$sample(10, 5)
  expect_type(x, "integer")
  expect_length(x, 5)
  expect_identical(length(unique(x)), 5L)
  expect_true(all(x >= 1))
  expect_true(all(x <= 10))
})

test_that("raw returns a raw vector", {
  rng <- randompack_rng("pcg64")
  x <- rng$raw(16)
  expect_type(x, "raw")
  expect_length(x, 16)
})

File: python/tests/test_discrete.py
import numpy as np

import randompack as rp


def pick_engine():
    engs = rp.engines()
    if "pcg64" in engs:
        return "pcg64"
    if "x256++simd" in engs:
        return "x256++simd"
    return next(iter(engs.keys()))


def test_int_vector_matrix():
    rng = rp.Rng(pick_engine())
    rng.seed(123)
    x = rng.int(-2, 3, size=50)
    assert np.all((-2 <= x) & (x <= 3))
    m = np.empty((6, 9), dtype=np.int64)
    rng.int(out=m, a=-4, b=-1)
    assert np.all((-4 <= m) & (m <= -1))


def test_perm():
    rng = rp.Rng(pick_engine())
    rng.seed(123)
    p = rng.perm(10)
    assert p.shape == (10,)
    assert np.array_equal(np.sort(p), np.arange(1, 11, dtype=np.int32))
    q = np.empty(7, dtype=np.int32)
    rng.perm(7, out=q)
    assert np.array_equal(np.sort(q), np.arange(1, 8, dtype=np.int32))


def test_sample():
    rng = rp.Rng(pick_engine())
    rng.seed(123)
    s = rng.sample(10, 5)
    assert s.shape == (5,)
    assert np.all((1 <= s) & (s <= 10))
    assert np.unique(s).size == 5
    t = np.empty(6, dtype=np.int32)
    rng.sample(12, 6, out=t)
    assert np.all((1 <= t) & (t <= 12))
    assert np.unique(t).size == 6


def test_raw():
    rng = rp.Rng(pick_engine())
    rng.seed(123)
    x = rng.raw(16)
    assert isinstance(x, (bytes,))
    assert len(x) == 16
